DOC Updated Nested GridField docs (#400)
symbiote/silverstripe-gridfieldextensions: docs/en/index.md

The `GridFieldNestedForm` component allows you to nest GridFields in the UI. It can be used with `DataObject` subclasses
with the `Hierarchy` extension, or by specifying the relation used for nesting.

Here is a small example of basic use of Nested GridField.

```php

> class MyAdmin extends ModelAdmin

Reviewer: Add `use` statements.
Please do this for all PHP code examples added in this PR.

Reviewer: Please also add a namespace for this and all other code examples added in this PR

Author: Updated

{

>     ...

Reviewer: ```suggestion
    //...
```

Author: Updated

    public function getEditForm($id = null, $fields = null)
    {
        $form = parent::getEditForm($id, $fields);

        $grid = $form->Fields()->dataFieldByName(MyObject::class);
        // Add Nested GridField to main GridField
        $grid->getConfig()->addComponent(GridFieldNestedForm::create());

        return $form;
    }
}
```

There are several ways to use Nested GridField. The implementation depends on the type of data you will be using in your GridField.

> For instance, if there is a DataObject that has the [`Hierarchy`](api:SilverStripe\ORM\Hierarchy\Hierarchy) extension, you can use the following approach.

Reviewer: ```suggestion
For instance, if there is a `DataObject` that has the [`Hierarchy`](api:SilverStripe\ORM\Hierarchy\Hierarchy) extension, you can use the following approach.
```

Author: Updated

> As an example, here we can use a typical hierarchy of Security Groups, where another Group can serve as a parent.

Reviewer: ```suggestion
As an example, here we can use a typical hierarchy of the `Group` model, where another `Group` can serve as a parent.
```

Author: Updated

```php
class MySecurityAdminExtension extends Extension
{
    public function updateGridFieldConfig($config)
    {
        if ($this->owner->getModelClass() === Group::class) {
            $config->addComponent(GridFieldNestedForm::create());
        }
    }
}
```

> Or also view a list of all members of a given group.

Reviewer: ```suggestion
Or also view a list of all members of a given group. Notice we call `setRelationName()` to explicitly state the relation which should be displayed in the Nested GridField.
```

Author: Updated

```php
class MySecurityAdminExtension extends Extension
{
    public function updateGridFieldConfig($config)
    {
        if ($this->owner->getModelClass() === Group::class) {
            $config->addComponent(GridFieldNestedForm::create()->setRelationName('Members'));
        }
    }
}
```

```yml
SilverStripe\Admin\SecurityAdmin:
  extensions:
    - MyApp\MyExtensions\MySecurityAdminExtension
```

Another way to use Nested GridField together with DataObjects that do not have the `Hierarchy` extension but have `has_many` relationships with other objects.

> Let's say you have the following DataObject that has multiple levels of relationships, and an Admin section where the data of this object will be displayed.

Reviewer: ```suggestion
Let's say you have the following `DataObject` that has multiple levels of relationships, and an admin section where the data of this object will be displayed.
```

Author: Updated

You want the user to be able to view information regarding this object and all nested objects on the same page, without the need to navigate to each object individually.
In this case, you can use the following approach.

> 

Author: I added a lot of code snippets below since it's not quite obvious which class should have Nested GridField.

```php
class ParentNode extends DataObject
{

>     ...

Reviewer: ```suggestion
    //...
```

Author: Updated

    private static $has_many = [
        'ChildNodes' => BranchNode::class,
    ];
}

> 

Reviewer: ```suggestion
```
Please remove the trailing new line in all of the new code blocks in this PR.

Author: Updated

```

```php
class ChildNode extends DataObject
{

>     ...

Reviewer: ```suggestion
    //...
```

Author: Updated

    private static $has_one = [
        'ParentNode' => ParentNode::class,
    ];

    private static $has_many = [
        'GrandChildNodes' => GrandChildNode::class,
    ];

    public function getNestedConfig(): GridFieldConfig
    {
        $config = new GridFieldConfig_RecordEditor();
        $config->addComponent(GridFieldNestedForm::create()->setRelationName('GrandChildNodes'));
        return $config;
    }
}

```

> 

Reviewer: Please add a paragraph here about the `getNestedConfig()` method.

Author: Updated

```php
class GrandChildNode extends DataObject
{

>     ...

Reviewer: ```suggestion
    //...
```

Author: Updated

    private static $has_one = [
        'ChildNode' => ChildNode::class,
    ];
}

```

```php
class MyAdminSection extends ModelAdmin
{
    private static string $url_segment = 'my-admin-section';
    private static string $menu_title = 'My Admin Section';

    private static array $managed_models = [
		ParentNode::class
	];

    protected function getGridFieldConfig(): GridFieldConfig
    {
        $config = parent::getGridFieldConfig();
        $config->addComponent(GridFieldNestedForm::create()->setRelationName('ChildNodes'));

        return $config;
    }
}

```

There is also the possibility to use Nested GridField with the data structure ArrayList. To do this, you can use the following approach.

```php
class MyDataSet extends DataObject
{

>     ...

Reviewer: ```suggestion
    //...
```

Author: Updated

    public function getMyArrayList() {
        $list = ArrayList::create();
        $data = MyDataObject::get();

        foreach ($data as $value) {
            $list->push($value);
        }

        return $list;
    }
}

```

```php
class MyAdminSection extends ModelAdmin
{

>     ...

Reviewer: ```suggestion
    //...
```

Author: Updated

    private static array $managed_models = [
		MyDataSet::class
	];

    protected function getGridFieldConfig(): GridFieldConfig
    {
        $config = parent::getGridFieldConfig();
        $config->addComponent(GridFieldNestedForm::create()->setRelationName('getMyArrayList'));

        return $config;
    }

> }

Author: Probably this test case should be updated with explanation how to set relation name, if user wants to show nested ArrayList in another ArrayList.

Reviewer: Yes please - even just a quick sentence saying something like this would be good:
> Notice that instead of the name of a relation, we're passing a method name into `getMyArrayList()`. That method must return an instance of `SS_List` to be used in the `GridField`.

Author: Updated

```

> #### Additional features and settings

Author: There are probably another cool features that should be mentioned in docs, but I could not find all of them.

1. You can set the maximum number of nested levels using a `$default_max_nesting_level` configuration property. The default value is 10 levels.

```yml
Symbiote\GridFieldExtensions\GridFieldNestedForm:
    default_max_nesting_level: 5

> ```

Reviewer: Please also add a quick example using the `setMaxNestingLevel()` method.

Author: Updated

> 1. You can set the maximum number of nested levels using a `$default_max_nesting_level` configuration property. The default value is 10 levels.
> 
> ```yml
> Symbiote\GridFieldExtensions\GridFieldNestedForm:
>     default_max_nesting_level: 5
> ```
> 
> 1. You can also modify the default config (a `GridFieldConfig_RecordEditor`) via an extension to the nested model class, by implementing
> `updateNestedConfig`, which will get the config object as the parameter.
> 

Reviewer: This isn't rendering correctly as a list. I think changing the indentation like this will work.
```suggestion
1. You can set the maximum number of nested levels using a `$default_max_nesting_level` configuration property. The default value is 10 levels.

    ```yml
    Symbiote\GridFieldExtensions\GridFieldNestedForm:
        default_max_nesting_level: 5
    ```

1. You can also modify the default config (a `GridFieldConfig_RecordEditor`) via an extension to the nested model class, by implementing
`updateNestedConfig`, which will get the config object as the parameter.

````

Author: Updated

use Symbiote\GridFieldExtensions\GridFieldNestedForm;

class MySecurityAdminExtension extends Extension
{
    public function updateGridFieldConfig($config)
    {
        if ($this->owner->getModelClass() === Group::class) {
            $config->addComponent(GridFieldNestedForm::create()->setRelationName('Members'));
        }
    }
}
```

```yml
SilverStripe\Admin\SecurityAdmin:
extensions:
    - App\Extensions\MySecurityAdminExtension
```

Another way to use Nested GridField together with `DataObjects` that do not have the `Hierarchy` extension but have `has_many` relationships with other objects.
Let's say you have the following `DataObject` that has multiple levels of relationships, and an admin section where the data of this object will be displayed.
You want the user to be able to view information regarding this object and all nested objects on the same page, without the need to navigate to each object individually.
In this case, you can use the following approach.

```php
namespace App\Models;

use SilverStripe\ORM\DataObject;

class ParentNode extends DataObject
{
    //...

    private static $has_many = [
        'ChildNodes' => BranchNode::class,
    ];
}
```

> You can define your own custom GridField config for the nested GridField configuration by implementing a `getNestedConfig` on your nested model. Notice this method should return a [`GridFieldConfig`](api:SilverStripe\Forms\GridField\GridFieldConfig) object.

Reviewer: ```suggestion
You can define your own custom GridField config for the nested GridField configuration by implementing a `getNestedConfig` on your nested model. Notice this method should return a `GridFieldConfig` object.
```
These docs aren't pulled into docs.silverstripe.org so we can't use that syntax unfortunately.

Author: Updated

> You can define your own custom GridField config for the nested GridField configuration by implementing a `getNestedConfig` on your nested model. Notice this method should return a [`GridFieldConfig`](api:SilverStripe\Forms\GridField\GridFieldConfig) object.

Reviewer: ```suggestion
You can define your own custom GridField config for the nested GridField configuration by implementing a `getNestedConfig()` method on your nested model. Notice this method should return a [`GridFieldConfig`](api:SilverStripe\Forms\GridField\GridFieldConfig) object.
```

Author: Updated

        return $config;
    }
}
```

There is also the possibility to use Nested GridField with the data structure ArrayList. To do this, you can use the following approach.

```php
namespace App\Models;

use MyDataObject;
use SilverStripe\ORM\ArrayList;
use SilverStripe\ORM\DataObject;

class MyDataSet extends DataObject
{
    //...

    public function getMyArrayList() {
        $list = ArrayList::create();
        $data = MyDataObject::get();

        foreach ($data as $value) {
            $list->push($value);
        }

        return $list;
    }
}
```

```php
namespace App\Admin;

use App\Models\MyDataSet;
use SilverStripe\Admin\ModelAdmin;
use SilverStripe\Forms\GridField\GridFieldConfig;
use Symbiote\GridFieldExtensions\GridFieldNestedForm;

class MyAdminSection extends ModelAdmin
{
    //...

    private static array $managed_models = [
        MyDataSet::class
    ];

    protected function getGridFieldConfig(): GridFieldConfig
    {
        $config = parent::getGridFieldConfig();
        $config->addComponent(GridFieldNestedForm::create()->setRelationName('getMyArrayList'));

        return $config;
    }
}
```

> Notice that instead of the name of a relation, we're passing a method name into `getMyArrayList()`. That method must return an instance of [`SS_List`](api:SilverStripe\ORM\SS_List) to be used in the GridField.

Reviewer: ```suggestion
Notice that instead of the name of a relation, we're passing a method name into `getMyArrayList()`. That method must return an instance of `SS_List` to be used in the GridField.
```

Author: Updated

#### Additional features and settings

1. You can set the maximum number of nested levels using a `$default_max_nesting_level` configuration property. The default value is 10 levels.

    ```yml
    Symbiote\GridFieldExtensions\GridFieldNestedForm:
        default_max_nesting_level: 5
    ```

>     You can also set this limitat only for a specific nested GridField using the `setMaxNestingLevel()` method.

Reviewer: ```suggestion
    You can also set this limit for a specific nested GridField using the `setMaxNestingLevel()` method.
```

Author: Updated
